File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> dp(amount+1, amount+1);
        dp[0] = 0;
        
        for (int i=1; i<amount+1; i++){
            for (int coin=0; coin<coins.size(); coin++){
                if (i >= coins[coin]){
                    dp[i] = min(dp[i-coins[coin]]+1, dp[i]);
                }
            }
        }
        if (dp[amount] == amount+1){
            return -1;
        }
        return dp[amount];
    }
};

File: 0322-coin-change/0322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [amount+1 for i in range(amount+1)]

        coins.sort()
        dp[0] = 0
        for i in range(amount+1):
            for j in range(len(coins)):
                coin = coins[j]
                if coin == i:
                    dp[i] = 1
                elif coin < i:
                    dp[i] = min(dp[i], 1 + dp[i-coin])
                else:
                    break
        
        return dp[amount] if dp[amount] != amount+1 else -1

